File: 221-MaximalSquare/221-MaximalSquare.java
// Last updated: 10/3/2025, 2:13:56 PM
class Solution {
    public int maximalSquare(char[][] matrix) {
        int maxSquare[][]=new int[matrix.length][matrix[0].length];

        for(int i=0;i<matrix.length;i++){
           for(int j=0;j<matrix[0].length;j++){
                  if(matrix[i][j]=='1')
                     maxSquare[i][j]=1;
           }
        }


        for(int i=1;i<matrix.length;i++){
            for(int j=1;j<matrix[0].length;j++){
                if(matrix[i][j]=='0') continue;
                maxSquare[i][j]+=Math.min(maxSquare[i-1][j-1],Math.min(maxSquare[i-1][j],maxSquare[i][j-1]));
            }   
        }
       int max=0;
        for(int i=0;i<matrix.length;i++){
            for(int j=0;j<matrix[0].length;j++){
                max=Math.max(max,maxSquare[i][j]);
            }
        }
        // System.out.println(Arrays.deepToString(maxSquare));
        // return (int)Math.pow(maxSquare[matrix.length-1][matrix[0].length-1],2);
        return (int)Math.pow(max,2);
    }
}

File: 221-MaximalSquare/221-MaximalSquare.py
# Last updated: 10/3/2025, 2:13:52 PM

class Solution(object):
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        m=len(matrix)
        n=len(matrix[0])
        mat=[[0]*n for _ in range(m)]
        for i in range (m):
            for j in range (n):
                if matrix[i][j]=='1':
                    mat[i][j]=1

        # print(mat)
        mac=0
        for i in range (1,m):
            for j in range(1,n):
                if(mat[i][j]==0): continue
                mat[i][j]+=min(mat[i-1][j],mat[i][j-1],mat[i-1][j-1])
                # mac=max(mac,mat[i][j])
        for i in range(m):
            for j in range(n):
                mac=max(mac,mat[i][j])
        return mac**2